internal/attestation/untyped.go

```diff
@@ -0,0 +1,161 @@
+const (
+	predicateTypeKey = "predicateType"
+	predicateKey     = "predicate"
+)
+
+// Similar function to in_toto.ProvenanceStatementSLSA02 but we'll use
+// an opaque untyped map
+type UntypedAttestation map[string]any
+
+// UntypedProvenanceFromSignature parses an attestation layer from the provided OCI
+// layer. Expects that the layer contains DSSE JSON with an embedded attestation
+// of some type or other.
+func UntypedProvenanceFromSignature(sig oci.Signature) (Attestation, error) {
+	payload, err := PayloadFromSig(sig)
+	if err != nil {
+		return nil, err
+	}
+
+	embedded, err := decodedPayload(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	var statement UntypedAttestation
+	if err := json.Unmarshal(embedded, &statement); err != nil {
+		return nil, AT002.CausedBy(err)
+	}
+
+	signatures, err := createEntitySignatures(sig, payload)
+	if err != nil {
+		return nil, AT005.CausedBy(err)
+	}
+
+	// Fix string predicate if needed
+	if statement.predicateIsString() {
+		fixedPredicate, err := statement.Predicate()
+		if err != nil {
+			return nil, AT002.CausedBy(err)
+		}
+		statement[predicateKey] = fixedPredicate
+	}
+
+	return untypedProvenance{statement: statement, data: embedded, signatures: signatures}, nil
+}
+
+// Detect if the value of the predicate key is a string
+func (att UntypedAttestation) predicateIsString() bool {
+	_, err := utils.MapGetStr(att, predicateKey)
+	return err == nil
+}
+
+// Attempt to extract the predicate from an arbitrary attestation, with a
+// workaround to parse a string predicate into an object if needed
+func (att UntypedAttestation) Predicate() (map[string]any, error) {
+	var predicate map[string]any
+
+	// Workaround what is (possibly) a bug in cosign where
+	// `cosign attest --predicate sbom-file.json` produces a
+	// json string version of the predicate instead of an unmarshaled
+	// object. Try to read a string first.
+	predicateStr, err := utils.MapGetStr(att, predicateKey)
+	if err == nil {
+		log.Debug("Found a string predicate. Unmarshalling it.")
+
+		// Found a string. Assume it's json and unmarshal it.
+		if err := json.Unmarshal([]byte(predicateStr), &predicate); err != nil {
+			return predicate, fmt.Errorf("Unable to unmarshal predicate string %q due to: %w", predicateStr, err)
+		}
+		return predicate, nil
+	}
+
+	// If it's not a string then it should be a map. We're expecting the format
+	// of the map could vary wildly depending on the predicateType
+	return utils.MapGetMap(att, predicateKey)
+}
+
+func (att UntypedAttestation) PredicateType() (string, error) {
+	return utils.MapGetStr(att, predicateTypeKey)
+}
+
+type untypedProvenance struct {
+	statement  UntypedAttestation
+	data       []byte
+	signatures []signature.EntitySignature
+}
+
+func (u untypedProvenance) Type() string {
+	// Maybe we don't know the predicateType, but it's still (presumably) an
+	// in-toto statement. Question: Should we extract it instead of hard coding it?
```

Comment:
Perhaps hardcoding this could allow someone to make a unsupported claim and trick the rules, perhaps, not sure. Smells a bit to have it hardcoded. Versions also change so it might be an issue in the future that would be a bit difficult to troubleshoot

Reply:
Yeah, I think we can probably revisit this question in future rather than decide on it now.